File: src/local_planner/paths/path_generator.m
clc;
clear all;
close all;

%% generate path
%{.
dis = 1.0; %%tune --> distance between x waypoints (increase; smoother decrease;  more reactive)
angle = 27; %%overall spread of the initial "V-shape"
deltaAngle = angle / 3; %%step size (smaller --> more paths generated)
scale = 0.65; %%reduces spread for deeper paths

pathStartAll = zeros(4, 0);
pathAll = zeros(5, 0);
pathList = zeros(5, 0);
pathID = 0;
groupID = 0;

figure;
hold on;
box on;
axis equal;
xlabel('X (m)');
ylabel('Y (m)');

%% Ackermann vehicle limits
wheelbase = 2.6;             % [m] TODO: measure
deltaMax = 30 * pi/180;      % [rad] maximum steering angle TODO: measure
kappaMax = tan(deltaMax) / wheelbase;   % maximum curvature

theta0 = 0;                   % vehicle initial heading along +X ?
maxYawChange = pi/2;          % max allowed heading change relative to current heading (±90°)
ds = 0.05;                    %step size: distance the vehicle moves forward along the path at each step.

function kappa = computeCurvature(x, y)
    dx  = gradient(x);
    dy  = gradient(y);
    ddx = gradient(dx);
    ddy = gradient(dy);
    kappa = abs(dx .* ddy - dy .* ddx) ./ ((dx.^2 + dy.^2).^(3/2));
end

fprintf('\nGenerating paths\n');

for shift1 = -angle : deltaAngle : angle
    wayptsStart = [0, 0, 0;
                   dis, shift1, 0];
    
    pathStartR = 0 : 0.01 : dis;
    pathStartShift = spline(wayptsStart(:, 1), wayptsStart(:, 2), pathStartR);
    
    pathStartX = pathStartR .* cos(pathStartShift * pi / 180);
    pathStartY = pathStartR .* sin(pathStartShift * pi / 180);
    pathStartZ = zeros(size(pathStartX));
    
    pathStart = [pathStartX; pathStartY; pathStartZ; ones(size(pathStartX)) * groupID];
    pathStartAll = [pathStartAll, pathStart];
    
    for shift2 = -angle * scale + shift1 : deltaAngle * scale : angle * scale + shift1
        for shift3 = -angle * scale^2 + shift2 : deltaAngle * scale^2 : angle * scale^2 + shift2
                waypts = [pathStartR', pathStartShift', pathStartZ';
                          2 * dis, shift2, 0;
                          3 * dis - 0.001, shift3, 0;
                          3 * dis, shift3, 0];

                pathR = 0 : 0: ds : waypts(end, 1);
                % %pathShift = spline(waypts(:, 1), waypts(:, 2), pathR);

                % %%prevent turning in place (alway moves forward now)
                % % spline lateral offset instead of heading
                % latOffset = spline(waypts(:,1), waypts(:,2), pathR);


                % % pathX = pathR .* cos(pathShift * pi / 180);
                % % pathY = pathR .* sin(pathShift * pi / 180);

                % %theta = pathShift * pi/180; 
                % % compute heading from geometry
                % dLat = gradient(latOffset, pathR);
                % theta = atan(dLat);     % small-angle Ackermann approximation

                % pathX = zeros(size(theta));
                % %pathY = zeros(size(theta));
                % pathY = latOffset;   % actual lateral offset


                % pathZ = zeros(size(pathX));
                % ds = 0.01; % step size = pathR step

                % for k = 2:length(theta)
                %     dx = pathR(k) - pathR(k-1);
                %     pathX(k) = pathX(k-1) + dx * cos(theta(k));
                %     pathY(k) = pathY(k-1) + dx * sin(theta(k));
                % end

                % %apply curvature check
                % kappa = computeCurvature(pathX, pathY);
                % if max(kappa) > kappaMax
                %     continue; %too sharp for ackemann steering --> skip the path
                % end
                 % Frenet lateral offset spline
            latOffset = spline(waypts(:,1), waypts(:,2), min(pathR, waypts(end,1)));
            dLat = gradient(latOffset, pathR);
            ddLat = gradient(dLat, pathR);
            
            % enforce curvature limits
            scaleFactor = 1.0;
            for iter=1:12
                kappaTest = abs(scaleFactor*ddLat ./ (1+(scaleFactor*dLat).^2).^(3/2));
                if max(kappaTest)<=kappaMax*1.0001
                    break;
                end
                scaleFactor = scaleFactor * 0.75;
            end
            latOffset = latOffset * scaleFactor;
            dLat = dLat * scaleFactor;
            ddLat = ddLat * scaleFactor;
            
            theta = atan(dLat); theta = unwrap(theta);
            kappa = ddLat ./ (1+dLat.^2).^(3/2);
            
         
            % prevent backwards motion
            pathX = zeros(size(theta));
            pathY = zeros(size(theta));
            pathY(1) = latOffset(1);
            for k=2:length(theta)
                dx = ds * cos(theta(k-1));
                if dx <= 0
                    dx = eps; % small positive step to prevent backward motion
                end
                pathX(k) = pathX(k-1) + dx;
                pathY(k) = pathY(k-1) + ds * sin(theta(k-1));
            end
            
            % U-turn (if final x < 0) %%now just drives backwards??
            if pathX(end) < 0
                extendFactorMax = 4; % max extension
                for ef = 2:extendFactorMax
                    newMaxS = ef * waypts(end,1);
                    pathR_new = 0:ds:newMaxS;
                    latOffset_new = spline(waypts(:,1), waypts(:,2), min(pathR_new, waypts(end,1)));
                    dLat_new = gradient(latOffset_new, pathR_new);
                    ddLat_new = gradient(dLat_new, pathR_new);
                    theta_new = atan(dLat_new); theta_new = unwrap(theta_new);

                    % forward-only integration
                    pathX_new = zeros(size(theta_new));
                    pathY_new = zeros(size(theta_new));
                    pathY_new(1) = latOffset_new(1);
                    for k2=2:length(theta_new)
                        dx = ds * cos(theta_new(k2-1));
                        if dx <= 0, dx = eps; end
                        pathX_new(k2) = pathX_new(k2-1) + dx;
                        pathY_new(k2) = pathY_new(k2-1) + ds * sin(theta_new(k2-1));
                    end

                    % check heading & curvature
                    pathAngle = atan2(pathY_new(end), pathX_new(end));
                    angleDiff = mod(pathAngle - theta0 + pi, 2*pi) - pi;
                    kappa_new = ddLat_new ./ (1 + dLat_new.^2).^(3/2);

                    if max(abs(kappa_new)) <= kappaMax*1.0001 && pathX_new(end) >= 0 && abs(angleDiff) <= maxYawChange
                        pathX = pathX_new; pathY = pathY_new; theta = theta_new; kappa = kappa_new;
                        break;
                    end
                end
            end

            % Final curvature check
            if max(abs(kappa)) > kappaMax
                    continue; % skip infeasible paths
            end

                
            % ---------------------------
            % Save path
            % ---------------------------
            pathZ = zeros(size(pathX));
            path = [pathX; pathY; pathZ; ones(size(pathX))*pathID; ones(size(pathX))*groupID];
            pathAll = [pathAll, path];
            pathList = [pathList, [pathX(end); pathY(end); pathZ(end); pathID; groupID]];
            pathID = pathID + 1;
            
            plot3(pathX, pathY, pathZ);
        end
    end
    
    groupID = groupID + 1
end

pathID

fileID = fopen('startPaths.ply', 'w');
fprintf(fileID, 'ply\n');
fprintf(fileID, 'format ascii 1.0\n');
fprintf(fileID, 'element vertex %d\n', size(pathStartAll, 2));
fprintf(fileID, 'property float x\n');
fprintf(fileID, 'property float y\n');
fprintf(fileID, 'property float z\n');
fprintf(fileID, 'property int group_id\n');
fprintf(fileID, 'end_header\n');
fprintf(fileID, '%f %f %f %d\n', pathStartAll);
fclose(fileID);

fileID = fopen('paths.ply', 'w');
fprintf(fileID, 'ply\n');
fprintf(fileID, 'format ascii 1.0\n');
fprintf(fileID, 'element vertex %d\n', size(pathAll, 2));
fprintf(fileID, 'property float x\n');
fprintf(fileID, 'property float y\n');
fprintf(fileID, 'property float z\n');
fprintf(fileID, 'property int path_id\n');
fprintf(fileID, 'property int group_id\n');
fprintf(fileID, 'end_header\n');
fprintf(fileID, '%f %f %f %d %d\n', pathAll);
fclose(fileID);

fileID = fopen('pathList.ply', 'w');
fprintf(fileID, 'ply\n');
fprintf(fileID, 'format ascii 1.0\n');
fprintf(fileID, 'element vertex %d\n', size(pathList, 2));
fprintf(fileID, 'property float end_x\n');
fprintf(fileID, 'property float end_y\n');
fprintf(fileID, 'property float end_z\n');
fprintf(fileID, 'property int path_id\n');
fprintf(fileID, 'property int group_id\n');
fprintf(fileID, 'end_header\n');
fprintf(fileID, '%f %f %f %d %d\n', pathList);
fclose(fileID);

pause(1.0);
%}

%% find correspondence
%{.
voxelSize = 0.02;
searchRadius = 0.45;
offsetX = 3.2;
offsetY = 4.5;
voxelNumX = 161;
voxelNumY = 451;

fprintf('\nPreparing voxels\n');

indPoint = 1;
voxelPointNum = voxelNumX * voxelNumY;
voxelPoints = zeros(voxelPointNum, 2);
for indX = 0 : voxelNumX - 1
    x = offsetX - voxelSize * indX;
    scaleY = x / offsetX + searchRadius / offsetY * (offsetX - x) / offsetX;
    for indY = 0 : voxelNumY - 1
        y = scaleY * (offsetY - voxelSize * indY);

        voxelPoints(indPoint, 1) = x;
        voxelPoints(indPoint, 2) = y;
        
        indPoint  = indPoint + 1;
    end
end

plot3(voxelPoints(:, 1), voxelPoints(:, 2), zeros(voxelPointNum, 1), 'k.');
pause(1.0);

fprintf('\nCollision checking\n');

[ind, dis] = rangesearch(pathAll(1 : 2, :)', voxelPoints, searchRadius);

fprintf('\nSaving correspondences\n');

fileID = fopen('correspondences.txt', 'w');

for i = 1 : voxelPointNum
    fprintf(fileID, '%d ', i - 1);
    
    indVoxel = sort(ind{i});
    indVoxelNum = size(indVoxel, 2);
    
    pathIndRec = -1;
    for j = 1 : indVoxelNum
        pathInd = pathAll(4, indVoxel(j));
        if pathInd == pathIndRec
            continue;
        end

        fprintf(fileID, '%d ', pathInd);
        pathIndRec = pathInd;
    end
    fprintf(fileID, '-1\n');
    
    if mod(i, 1000) == 0
        i
    end
end

fclose(fileID);

fprintf('\nProcessing complete\n');
%}

File: src/local_planner/paths/path_generator.py
import numpy as np
from scipy.interpolate import CubicSpline
import os
os.chdir(os.path.dirname(os.path.abspath(__file__)))


dis = 1
angle = 90
deltaAngle = angle / 3
scale = 0.1

pathStartAll = []
pathAll = []
pathList = []
pathID = 0
groupID = 0

# Ackermann limits
wheelbase = 0.3
deltaMax = 30 * np.pi / 180
kappaMax = np.tan(deltaMax) / wheelbase

theta0 = 0.0
maxYawChange = np.pi / 2
ds = 0.05


# utility functions
def compute_curvature(x, y):
    dx = np.gradient(x)
    dy = np.gradient(y)
    ddx = np.gradient(dx)
    ddy = np.gradient(dy)
    kappa = np.abs(dx * ddy - dy * ddx) / (dx**2 + dy**2)**1.5
    kappa[np.isnan(kappa)] = 0
    return kappa


def spline(x, y, xq):
    """1D spline interpolation"""
    cs = CubicSpline(x, y, extrapolate=True)
    return cs(xq)


# generating paths
print("\nGenerating paths")

for shift1 in np.arange(-angle, angle + 1e-6, deltaAngle):
    wayptsStart = np.array([[0, 0, 0],
                            [dis, shift1, 0]])

    pathStartR = np.arange(0, dis + 1e-6, 0.01)
    pathStartShift = spline(wayptsStart[:, 0], wayptsStart[:, 1], pathStartR)

    pathStartX = pathStartR * np.cos(pathStartShift * np.pi / 180)
    pathStartY = pathStartR * np.sin(pathStartShift * np.pi / 180)
    pathStartZ = np.zeros_like(pathStartX)

    pathStart = np.vstack([pathStartX, pathStartY, pathStartZ,
                            np.ones_like(pathStartX) * groupID])
    pathStartAll.append(pathStart)

    for shift2 in np.arange(-angle * scale + shift1,
                            angle * scale + shift1 + 1e-6,
                            deltaAngle * scale):
        for shift3 in np.arange(-angle * scale**2 + shift2,
                                angle * scale**2 + shift2 + 1e-6,
                                deltaAngle * scale**2):

            waypts = np.array([
                [pathStartR[-1], pathStartShift[-1], 0],
                [2 * dis, shift2, 0],
                [3 * dis - 0.001, shift3, 0],
                [3 * dis, shift3, 0]
            ])

            pathR = np.arange(0, waypts[-1, 0] + 1e-6, ds)

            # frenet lateral offset spline
            latOffset = spline(waypts[:, 0], waypts[:, 1],
                               np.minimum(pathR, waypts[-1, 0]))
            dLat = np.gradient(latOffset, pathR)
            ddLat = np.gradient(dLat, pathR)

            # enforce curvature limits
            scaleFactor = 1.0
            for _ in range(12):
                kappaTest = np.abs(scaleFactor * ddLat /
                                   (1 + (scaleFactor * dLat)**2)**1.5)
                if np.max(kappaTest) <= kappaMax * 1.0001:
                    break
                scaleFactor *= 0.75

            latOffset *= scaleFactor
            dLat *= scaleFactor
            ddLat *= scaleFactor

            theta = np.unwrap(np.arctan(dLat))
            kappa = ddLat / (1 + dLat**2)**1.5

            # prevent backwards motion
            pathX = np.zeros_like(theta)
            pathY = np.zeros_like(theta)
            pathY[0] = latOffset[0]

            for k in range(1, len(theta)):
                dx = ds * np.cos(theta[k - 1])
                if dx <= 0:
                    dx = np.finfo(float).eps
                pathX[k] = pathX[k - 1] + dx
                pathY[k] = pathY[k - 1] + ds * np.sin(theta[k - 1])

            # U-turn handling
            if pathX[-1] < 0:
                extendFactorMax = 4
                for ef in range(2, extendFactorMax + 1):
                    newMaxS = ef * waypts[-1, 0]
                    pathR_new = np.arange(0, newMaxS + 1e-6, ds)
                    latOffset_new = spline(waypts[:, 0], waypts[:, 1],
                                            np.minimum(pathR_new, waypts[-1, 0]))
                    dLat_new = np.gradient(latOffset_new, pathR_new)
                    ddLat_new = np.gradient(dLat_new, pathR_new)
                    theta_new = np.unwrap(np.arctan(dLat_new))

                    pathX_new = np.zeros_like(theta_new)
                    pathY_new = np.zeros_like(theta_new)
                    pathY_new[0] = latOffset_new[0]

                    for k2 in range(1, len(theta_new)):
                        dx = ds * np.cos(theta_new[k2 - 1])
                        if dx <= 0:
                            dx = np.finfo(float).eps
                        pathX_new[k2] = pathX_new[k2 - 1] + dx
                        pathY_new[k2] = pathY_new[k2 - 1] + ds * np.sin(theta_new[k2 - 1])

                    pathAngle = np.arctan2(pathY_new[-1], pathX_new[-1])
                    angleDiff = (pathAngle - theta0 + np.pi) % (2 * np.pi) - np.pi
                    kappa_new = ddLat_new / (1 + dLat_new**2)**1.5

                    if (np.max(np.abs(kappa_new)) <= kappaMax * 1.0001 and
                        pathX_new[-1] >= 0 and
                        abs(angleDiff) <= maxYawChange):
                        pathX, pathY, theta, kappa = pathX_new, pathY_new, theta_new, kappa_new
                        break

            # final curvature check
            if np.max(np.abs(kappa)) > kappaMax:
                continue

            # save path
            pathZ = np.zeros_like(pathX)
            path = np.vstack([pathX, pathY, pathZ,
                              np.ones_like(pathX) * pathID,
                              np.ones_like(pathX) * groupID])
            pathAll.append(path)
            pathList.append([pathX[-1], pathY[-1], pathZ[-1], pathID, groupID])
            pathID += 1

    groupID += 1
    print("groupID =", groupID)

print("Total paths:", pathID)


# save paths as PLY
def save_ply(filename, data, properties):
    with open(filename, 'w') as f:
        f.write("ply\n")
        f.write("format ascii 1.0\n")
        f.write(f"element vertex {data.shape[1]}\n")
        for p in properties:
            f.write(p + "\n")
        f.write("end_header\n")
        fmt = " ".join(["%f"] * (data.shape[0] - 2) + ["%d", "%d"])
        np.savetxt(f, data.T, fmt=fmt)


pathStartAll_arr = np.hstack(pathStartAll)
pathAll_arr = np.hstack(pathAll)
pathList_arr = np.array(pathList).T

save_ply("startPaths.ply", pathStartAll_arr,
         ["property float x",
          "property float y",
          "property float z",
          "property int group_id"])

save_ply("paths.ply", pathAll_arr,
         ["property float x",
          "property float y",
          "property float z",
          "property int path_id",
          "property int group_id"])

save_ply("pathList.ply", pathList_arr,
         ["property float end_x",
          "property float end_y",
          "property float end_z",
          "property int path_id",
          "property int group_id"])

print("\nPLY files written.")